src/nvim/lua/treesitter.c

```diff
@@ -391,10 +396,12 @@ static int parser_parse(lua_State *L)
     return luaL_argerror(L, 3, "expected either string or buffer handle");
   }
 
-  // Sometimes parsing fails (timeout, or wrong parser ABI)
-  // In those case, just return an error.
+  // As we ensure that the parser ABIs are correct when creating the languages,
+  // we are sure that the only reason of receiving a NULL is because of a
+  // timeout.
   if (!new_tree) {
-    return luaL_error(L, "An error occurred when parsing.");
+    ts_parser_reset(*p);
```

Comment:
Do we really want to reset the parser unconditionally? IIRC this is only necessary if there is an edit after an incomplete parse was aborted. otherwise, next time parsing the buffer is attempted, we should just let it continue on the existing job (still having the same deadline of course).

Reply:
Hmm maybe that's better indeed... I'll try to handle that then.